src/libs/OptionsListUtils.js

```diff
@@ -587,8 +588,8 @@ function getOptions(reports, personalDetails, activeReportID, {
 
     let userToInvite = null;
     if (searchValue
-        && recentReportOptions.length === 0
-        && personalDetailsOptions.length === 0
+        && ((recentReportOptions.length + personalDetailsOptions.length) === 0
+        || (!_.find(loginOptionsToExclude, loginOptionToExclude => loginOptionToExclude.login === searchValue.toLowerCase())))
```

Comment:
**thought:** It might help readability if we just extract this whole if block to a method like `canAddUserToInvite()`? It looks pretty crazy and we are duplicating some logic for the login options.

Reply:
I have this feeling too.

What do you think of changing the if block to:
```js
if (canAddUserToInvite(searchValue,recentReportOptions,personalDetailsOptions,loginOptionsToExclude,selectedOptions,loginOptionsToExclude))
```
adding the method:
```js
function canAddUserToInvite(searchValue,recentReportOptions,personalDetailsOptions,loginOptionsToExclude,selectedOptions){
    const canAddUserToInvite = searchValue && ((recentReportOptions.length + personalDetailsOptions.length) === 0
    || (!_.find(loginOptionsToExclude, loginOptionToExclude => loginOptionToExclude.login === searchValue.toLowerCase())))
    && !isCurrentUser({login: searchValue})
    && _.every(selectedOptions, option => option.login !== searchValue)
    && ((Str.isValidEmail(searchValue) && !Str.isDomainEmail(searchValue)) || Str.isValidPhone(searchValue))
    && (!_.find(loginOptionsToExclude, loginOptionToExclude => loginOptionToExclude.login === searchValue.toLowerCase()))

    return canAddUserToInvite;
}
```

Reply:
Yeah that's a lot of variables for a method, but cleaning it up would be an improvement. We could also maybe making it less of a giant ternary that is hard to make sense out of and use extra variables and if statements (but let's pause that for a second).

As for the redundancy, I agree with the assessment that `B` would be unneeded - but is the proposal to remove the code to check for available options? Why was the change added in the first place?

Reply:
this change was added in order to prevent this edge case @Santhosh-Sellavel found on https://github.com/Expensify/App/issues/8007#issuecomment-1064311667 , so we needed to show user to invite even when it doesn't found a result from search(when lenghts == 0)

Reply:
Maybe we could just break down conditions into variables with proper naming & comments if needed. 
Not sure about adding a new method `canAddUserToInvite`, it feels unnecessary.

Reply:
Sorry can you just explain that a bit more? I'm not following what the edge case is or how the change addressed it.

Reply:
>break down conditions into variables with proper naming & comments if needed

Yes let's do this? This logic is hard to follow especially with the new condition (we are improving it?)

Reply:
@marcaaron with the logic we have at the moment we will show the `user to invite` when we don't find results from search.
e.g.:
If I search `marcaaron@expensify.com`
I don't have a report with `marcaaron@expensify.com`, so it will have 0 results and will show `marcaaron@expensify.com` as an `user to invite`.

Now lets suppose I have a report with `marcaaron@expensify.com` and I want to create a report with `marc.aaron@expensify.com` (we are dealing with two differents emails from two differents accounts).

Searching `marc.aaron@expensify.com`(with our changes) will result on finding `marcaaron@expensify.com` and we will have , `(recentReportOptions.length + personalDetailsOptions.length) > 0`, so we won't show `marc.aaron@expensify.com` as an `user to invite` when we should.

Reply:
>I want to create a report with marc.aaron@expensify.com (we are dealing with two differents emails from two differents accounts).

Ah hmm well not if you have a gmail account then the [emails would be the same](https://slate.com/technology/2013/08/dots-in-gmail-addresses-what-happens-if-you-leave-out-the-period.html) 😄 

I guess it would be hard (or maybe pointless) to add logic to try change things based on the email provider so I agree that if you type in a unique email it should bring up the user to invite with the dots.

Sounds like the logic we need should just be:

- Show all possible options that loosely match with or without periods
- If our `searchValue` itself is a valid email or phone and it does not **exactly** match ANY of the available options then it should be shown as a user to invite

What I'm confused about is why we're looking at `loginOptionsToExclude`.

Shouldn't the logic be more like:

```js
const noOptions = recentReportOptions.length + personalDetailsOptions.length) === 0;
const noOptionsMatchExactly = // whatever function that will check the search value is a valid login and that none of the options match exactly
if (searchValue && (noOptions || noOptionsMatchExactly) && ...
```

Reply:
```js
const noOptions = recentReportOptions.length + personalDetailsOptions.length) === 0;
const noOptionsMatchExactly = // whatever function that will check the search value is a valid login and that none of the options match exactly
if (searchValue && (noOptions || noOptionsMatchExactly) && ...
```
Great @marcaaron I couldn’t write anything better.

> Ah hmm well not if you have a Gmail account then the [emails would be the same](https://slate.com/technology/2013/08/dots-in-gmail-addresses-what-happens-if-you-leave-out-the-period.html) 😄

We would allow creating accounts with the same Gmail account with different dot patterns, also wondering who would create accounts like this?

Reply:
Yeah, it's possible, but probably quite rare. 

**Example:** `marc@gmail.com` and `ma.rc@gmail.com` can be created as two separate Expensify accounts. Even though any email sent to `ma.rc@gmail.com` will go to `marc@gmail.com` on the mail server side.

>who would create accounts like this?

Maybe nobody ever or if they did it would likely be an accident. I think this is an extreme edge case, but since it's possible to create two Expensify accounts with dots in the address then we should treat them as different. So...

- If the searchValue is a valid email and does **not** appear in the list of filtered options **exactly** then it can be added
- If the searchValue is not a valid email or **does** appear in the list then we show the various options

Reply:
So, I think we reached this:
```js
    const noOptions = (recentReportOptions.length + personalDetailsOptions.length) === 0;
    const noOptionsMatchExactly = !_.find(loginOptionsToExclude, loginOptionToExclude => loginOptionToExclude.login === searchValue.toLowerCase());// whatever function that will check the search value is a valid login and that none of the options match exactly
    if (searchValue && (noOptions || noOptionsMatchExactly)
        && !isCurrentUser({login: searchValue})
        && _.every(selectedOptions, option => option.login !== searchValue)
        && ((Str.isValidEmail(searchValue) && !Str.isDomainEmail(searchValue)) || Str.isValidPhone(searchValue))
        && noOptionsMatchExactly
        && (searchValue !== CONST.EMAIL.CHRONOS || Permissions.canUseChronos(betas))
```
and by https://github.com/Expensify/App/pull/8375#discussion_r844355317 I think we can simplify to this:
```js
    const noOptionsMatchExactly = !_.find(loginOptionsToExclude, loginOptionToExclude => loginOptionToExclude.login === searchValue.toLowerCase());// whatever function that will check the search value is a valid login and that none of the options match exactly
    if (searchValue
        && !isCurrentUser({login: searchValue})
        && _.every(selectedOptions, option => option.login !== searchValue)
        && ((Str.isValidEmail(searchValue) && !Str.isDomainEmail(searchValue)) || Str.isValidPhone(searchValue))
        && noOptionsMatchExactly
        && (searchValue !== CONST.EMAIL.CHRONOS || Permissions.canUseChronos(betas))
```
and we will keep the same behavior.
do you guys agree with that?

If so, I'll commit that.

Reply:
Sorry @mateusbra I am still stuck on this:

>What I'm confused about is why we're looking at `loginOptionsToExclude`. 

Why are we looking at the excluded options list and not the combined list of options?

Reply:
I think we need to use the combined results of both `recentReportOptions` and `personalDetailsOptions`. IIRC both can have a `login` value and the person we are looking for could be either a "report participant" or a "personal detail" (meaning we do not have a chat with them yet but might have some other association with them).

Reply:
So I think we should add to `noOptionsMatchExactly` this condition, so we get both `recentReportOptions` and `personalDetailsOptions`:
```js
  const noOptionsMatchExactly = !_.find(personalDetailsOptions, personalDetailsOption => personalDetailsOption.login === searchValue.toLowerCase()) 
    || !_.find(recentReportOptions, recentReportOption => recentReportOption.login === searchValue.toLowerCase());
```
what you think?

Reply:
> What I'm confused about is why we're looking at loginOptionsToExclude.

@marcaaron this was made to exclude 

- preselected options
- excluded Expensify Logins

Reply:
@Santhosh-Sellavel Sorry, I don't understand. Can you explain in detail why that suggestion was made or verify that it was incorrect?

@mateusbra better but we can maybe simplify that to

```js
!_.find(personalDetailsOptions.concat(recentReportOptions), option => option.login === searchValue.toLowerCase());
```

Reply:
@marcaaron nice one! may I commit all thoses changes?

Reply:
Yep!

Reply:
Sorry, @marcaaron for the confusion. 

You are right,
> looking at `loginOptionsToExclude` doesn’t make sense.

The confusion originated from my suggestions here. https://github.com/Expensify/App/pull/8375#issuecomment-1084997941 those were two quick suggestions.

But we would really avoided this conversation if @mateusbra went with first suggestion, because this occurs only when search has dot.
 ```js
 && ((recentReportOptions.length + personalDetailsOptions.length) === 0 || _.contains(searchValue, '.'))
 ```
 
 But this is the better & simpler solution, unfortunately we both missed to foresee
 ```js
 !_.find(personalDetailsOptions.concat(recentReportOptions), option => option.login === searchValue.toLowerCase());
 ```

Reply:
Sorry @Santhosh-Sellavel I thought the second sugestion you proposed was better 😅, but now I think the misunderstoods were solved.

Reply:
I commited the changes, if you think its better to use the first suggestion from https://github.com/Expensify/App/pull/8375#issuecomment-1084997941 please tell me.